TypeCobol/Compiler/CupParser/NodeBuilder/ProgramClassBuilder.cs

```diff
@@ -230,12 +231,48 @@ public virtual void StartCobolProgram(ProgramIdentification programIdentificatio
             Dispatcher.StartCobolProgram(programIdentification, libraryCopy);
         }
 
-        public virtual void EndCobolProgram(TypeCobol.Compiler.CodeElements.ProgramEnd end)
+        public virtual void EndCobolProgram(ProgramEnd end)
         {
-            AttachEndIfExists(end);
-            Exit();
-            programsStack.Pop();
-            Dispatcher.EndCobolProgram(end);
+            if (end != null)
+            {
+                if (end.ProgramName?.Name != null)
+                {
+                    var matchingProgram = FindMatchingProgram();
+                    if (matchingProgram != null)
+                    {
+                        // Pop all programs to have CurrentProgram as the program whose name match
+                        while (matchingProgram != CurrentProgram)
+                        {
+                            // Skip closing all nested programs
+                            Exit();
```

Comment:
Don't forget to call the Dispatcher

Reply:
I tried the recursive calls and it seems to work better that way. There is no more need to do the strange test as you said.